designs/degraded-nodepools.md

```diff
@@ -0,0 +1,107 @@
+Status:
+  Conditions:
+    Last Transition Time:  2025-01-13T18:57:20Z
+    Message:               
+    Observed Generation:   1
+    Reason:                Degraded
+    Status:                True
+    Type:                  Degraded
+```
+`Degraded` status condition is introduced in the NodePool status which can be set to - 
+1. Unknown - When the NodePool is first created, `Degraded` is set to Unknown. This means that we don't have enough data to tell if the NodePool is degraded or not. 
+2. True - NodePool has configuration issues that require customer investigation and resolution. Since Karpenter cannot automatically detect these specific launch or registration failures, we will document common failure scenarios and possible fixes in our troubleshooting guide to assist customers.
+3. False - There has been successful node registration using this NodePool.
+
+The state transition is not unidirectional meaning it can go from True to False and back to True or Unknown. A NodePool marked as Degraded can still be used for provisioning workloads, as this status isn't a precondition for readiness. However, when multiple NodePools have the same weight, a degraded NodePool will receive lower priority during the provisioning process compared to non-degraded ones. 
+
+The approach that we go forward with should -
+1. Tolerate transient errors.
+2. Respond to corrections in external configuration (i.e. can remove the degraded status condition from a NodePool if an external fix allows Nodes to register).
+
+### Option 1: In-memory Buffer to store history - Recommended
+
+This option will have an in-memory FIFO buffer, which will grow to a max size of 10 (this can be changed later). This buffer will store data about the success or failure during launch/registration and is evaluated by a controller to determine the relative health of the NodePool. This will be an int buffer and a positive means `Degraded: False`, negative means `Degraded: True` and 0 means `Degraded: Unknown`.
+
+![](./images/degraded-nodepools1.png)
+
+Evaluation conditions -
+
+1. We start with an empty buffer with `Degraded: Unknown`.
+2. There have to be 2 minimum failures in the buffer for `Degraded` to transition to `True`. 
+3. If the buffer starts with a success then `Degraded: False`. 
+4. If Karpenter restarts then we flush the buffer but don't change the existing state of `Degraded` status condition.
+5. If there is an update to a Nodepool/Nodeclass, flush the buffer and set `Degraded: Unknown`.
+6. Since the buffer is FIFO, we remove the oldest launch result when the max buffer size is reached.
+
+See below for example evaluations:
+
+```
+Successful Launch: 1
+Default: 0
+Unsuccessful Launch: -1
+
+[] = 'Degraded: Unknown'
+[+1] = 'Degraded: False'
+[-1, +1] = 'Degraded: False'
+[-1, -1] = 'Degraded: True'
+[-1, +1, -1] = 'Degraded: True'
+[-1, +1, +1, +1, +1, +1, +1, +1, +1, +1] = 'Degraded: False'
+```
+
+#### Considerations
```

Comment:
We discussed this but what happens if we have one success and that causes us to stop trying new NodeClaims -- when that happens, what's the way that we make sure that we eventually get out of a Degraded state

Reply:
I haven't updated the RFC since our discussion about that. But @rschalo and I were thinking about expiring the entries in the buffer after some time - maybe 3x the registration ttl.

Reply:
I think one of the options we had discussed was expiring entries some amount of time after the last write so that there is some recency bias.

Reply:
Expiring the entries like this should also see when was the last update made to the buffer if it is frequent enough (we can define the time), then we don't expire entries until the buffer is full.